--- IMG_png.c.orig	2011-09-16 21:06:59 +0800
+++ IMG_png.c	2011-09-16 21:10:31 +0800
@@ -70,6 +70,8 @@
 #define MACOS
 #endif
 #include <png.h>
+#include <pngstruct.h>
+#include <pnginfo.h>
 
 
 static struct {
@@ -347,7 +349,7 @@
 	 * the normal method of doing things with libpng).  REQUIRED unless you
 	 * set up your own error handlers in png_create_read_struct() earlier.
 	 */
-	if ( setjmp(png_ptr->jmpbuf) ) {
+	if ( setjmp(png_ptr->longjmp_buffer) ) {
 		error = "Error reading the PNG file.";
 		goto done;
 	}
